### PR TITLE
fix: includes variant prices when listing products using a search query

### LINKED ISSUE #1484: List-products endpoint does not return prices when queried with search filter
After updating from v1.2.1 to v1.3

I'm no longer getting prices in the response of the list-products endpoint when I'm querying with a search filter.

This correctly returns prices -> `GET /products` 
While this does not -> `GET /products?q=text`

Is this expected behavior (given there have been some price selection strategy changes in recent version) ?
If yes, how to eager fetch the prices along with list of products.

### PR DESCRIPTION
**Pull request by zakariaelas**
### What

- Fixes #1484 

### Why the bug happens

The bug occurs because whenever there is a search query, we [return early](https://github.com/medusajs/medusa/blob/develop/packages/medusa/src/services/product.js#L194) from the `listAndCount` method causing the price calculation strategy, which only comes [later](https://github.com/medusajs/medusa/blob/develop/packages/medusa/src/services/product.js#L202-L213), to never run

### How

- add a `getFreeTextSearchResultsAndCount` method to the `ProductRepository`
- refactor `list` and `listAndCount` from the `ProductService` to use the `getFreeTextSearchResultsAndCount` method
- remove the `getFreeTextQueryBuilder_` method from the `ProductService` as we should aim to keep db related code away from the `ProductService`

### Testing

- add test case for making sure we get variant prices back when listing products using a search query
- enhance existing test case with extra check to avoid false positives

### CODE REVIEW

File: packages/medusa/src/repositories/product.ts [hunk 1]
```diff
@@ -314,4 +323,60 @@ export class ProductRepository extends Repository<Product> {
...
+    const [results, count] = await qb.getManyAndCount()
+
+    const products = await this.findWithRelations(
+      relations,
+      results.map((r) => r.id),
+      cleanedOptions.withDeleted
+    )
+
+    return [products, count]
+  }
```
Comment (olivermrbl) [inline packages/medusa/src/repositories/product.ts:364]: suggestion: We already have a very similar (almost identical) method in the product service [here](https://github.com/medusajs/medusa/blob/2eb2126fe348565721e05e6f0da01c46bad6185a/packages/medusa/src/services/product.js#L1014). Should we try to merge those two, or have you intentionally created a new one? 👍

File: packages/medusa/src/services/product.js [hunk 2]
```diff
@@ -182,23 +181,21 @@ class ProductService extends BaseService {
...
     const { q, query, relations } = this.prepareListQuery_(selector, config)
 
+    let products
+    let count
     if (q) {
-      const qb = this.getFreeTextQueryBuilder_(productRepo, query, q)
-      const [raw, count] = await qb.getManyAndCount()
-
-      const products = await productRepo.findWithRelations(
+      ;[products, count] = await productRepo.getFreeTextSearchResultsAndCount(
```
Comment (olivermrbl) [inline packages/medusa/src/services/product.js:187]: comment: Never seen syntax like this `;[variable]`. Where did you come across that and what does it do? 😅

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Super nice! Just have a couple of comments to address 👍

File: packages/medusa/src/repositories/product.ts [hunk 3]
```diff
@@ -314,4 +323,60 @@ export class ProductRepository extends Repository<Product> {
...
+    options: CustomOptions = { where: {} },
+    relations: (keyof Product)[] = []
+  ): Promise<[Product[], number]> {
+    const cleanedOptions = this._cleanOptions(options)
+
+    let qb = this.createQueryBuilder("product")
+      .leftJoinAndSelect("product.variants", "variant")
+      .leftJoinAndSelect("product.collection", "collection")
+      .select(["product.id"])
+      .where(cleanedOptions.where)
```
Comment (adrien2p) [inline packages/medusa/src/repositories/product.ts:338]: suggestion: You can move the content of the clean method here since it is pretty specific. as mention by oli this method already exists in the service :)

File: packages/medusa/src/repositories/product.ts [hunk 4]
```diff
@@ -314,4 +323,60 @@ export class ProductRepository extends Repository<Product> {
...
+    const products = await this.findWithRelations(
+      relations,
+      results.map((r) => r.id),
+      cleanedOptions.withDeleted
+    )
+
+    return [products, count]
+  }
+
+  private _cleanOptions(
```
Comment (adrien2p) [inline packages/medusa/src/repositories/product.ts:366]: comment: As suggested above, not sure that method should exists, at least at the moment and with that impl.

File: packages/medusa/src/services/product.js (hunk 2 above)
Comment (adrien2p) [reply packages/medusa/src/services/product.js:187]: does not do much, it would have been to separate two steps made on the same line. example:
```ts
doSomething()[0].forEach(...) // We take the first element of the returned array from doSomething
doSomething();[0].forEach(...) // Now they are two separate actions
// With the actual format
doSomething()
;[0].forEach(...)
```

Comment (adrien2p) [reply packages/medusa/src/services/product.js:187]: [suggestion: Remove the `;`

File: packages/medusa/src/services/product.js [hunk 5]
```diff
@@ -182,23 +181,21 @@ class ProductService extends BaseService {
...
+        q,
+        query,
+        relations
+      )
+    } else {
+      ;[products, count] = await productRepo.findWithRelationsAndCount(
```
Comment (adrien2p) [inline packages/medusa/src/services/product.js:193]: suggestion: Remove `;` should not be necessary

File: packages/medusa/src/types/utils.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Utility type used to remove some optional attributes (coming from K) from a type T
+ */
+export type WithRequiredProperty<T, K extends keyof T> = T &
```
Comment (adrien2p) [inline packages/medusa/src/types/utils.ts:4]: suggestion: Move that to common.ts since this the place where `PartialPick` and `Writable` are :)

**Review by adrien2p — CHANGES_REQUESTED** [review]
Nice Job 💪 few comments to address

File: packages/medusa/src/repositories/product.ts (hunk 1 above)
Comment (zakariaelas) [reply packages/medusa/src/repositories/product.ts:364]: @olivermrbl correct! It's pretty much the same thing, I just moved the function to the repo layer where I think it should belong as we've been trying to limit interacting with our persistence mechanism through the service layer (and instead do it through the repo layer) 😄

File: packages/medusa/src/services/product.js (hunk 2 above)
Comment (zakariaelas) [reply packages/medusa/src/services/product.js:187]: Very short [read](https://prettier.io/docs/en/rationale.html#semicolons) that I think will do a much better job at explaining that I will 😆 

TL;DR: it's added by prettier as a protective measure

File: packages/medusa/src/services/product.js (hunk 5 above)
Comment (zakariaelas) [reply packages/medusa/src/services/product.js:193]: addressed in previous comment (recap: this is added by prettier as a protective measure)

File: packages/medusa/src/services/product.js (hunk 2 above)
Comment (adrien2p) [reply packages/medusa/src/services/product.js:187]: Yes this is true, never thought that someone would miss it but hey we never know 😂

File: packages/medusa/src/services/product.js (hunk 5 above)
Comment (adrien2p) [reply packages/medusa/src/services/product.js:193]: Yes, means that having the debate on the end of line semilicon is more in favour of having them ^^

File: packages/medusa/src/repositories/product.ts (hunk 4 above)
Comment (zakariaelas) [reply packages/medusa/src/repositories/product.ts:366]: hmm .. I tried giving this some thought and I feel like there is not really a disadvantage with having the method in the repo class. In fact, one might even argue that it's better because reading the `getFreeTextSearchResultsAndCount` method becomes a tiny bit easier to read. When reading the method, you learn that we're cleaning the options without paying too much attention to the specifics of what it does (which can be a bit distracting if left inside the `getFreeTextSearchResultsAndCount` method). 

Ofc, I don't think this is too important and I'm okay with removing the method, just thought I'd share my thoughts here in case you have some counter arguments 😄

File: integration-tests/api/__tests__/admin/product.js [hunk 7]
```diff
@@ -256,6 +261,36 @@ describe("/admin/products", () => {
       expect(response.data.products.length).toEqual(2)
     })
 
+    it("returns a list of products with free text query including variant prices", async () => {
```
[the remark this answers is not in the document]
Comment (zakariaelas) [reply integration-tests/api/__tests__/admin/product.js:264]: good point!

File: packages/medusa/src/repositories/product.ts (hunk 4 above)
Comment (adrien2p) [reply packages/medusa/src/repositories/product.ts:366]: The thing is that it is equivalent at 
delete where.description 
delete where.title

which I think is still readable and self explanatory.
This is why i thought for the moment this method does not add a lot of value in term of
Separation nor understanding.

But in the end, i can be fine with both😂

File: packages/medusa/src/repositories/product.ts [hunk 8]
```diff
@@ -314,4 +323,60 @@ export class ProductRepository extends Repository<Product> {
...
+    options: CustomOptions
+  ): WithRequiredProperty<CustomOptions, "where"> {
+    const where = options.where ?? {}
+    if ("description" in where) {
```
Comment (adrien2p) [inline packages/medusa/src/repositories/product.ts:370]: suggestion: you can simplify that like this:
```ts
  const where = options.where ?? {}
  delete where.description
  delete where.title
  return { ...options, where }
```

**Review by adrien2p — APPROVED** [review]
LGTM 🚀 only few suggestions

**Review by olivermrbl — APPROVED** [review]
LGTM!

File: packages/medusa/src/repositories/product.ts (hunk 8 above)
Comment (zakariaelas) [reply packages/medusa/src/repositories/product.ts:370]: I tried that, but unfortunately, typescript is not too happy about that. This is the error I get:
```
Property 'description' does not exist on type '(string | ObjectLiteral | FindConditions<Product> | FindConditions<Product>[] | undefined) & { tags?: FindOperator<ProductTag> | undefined; price_list_id?: FindOperator<...> | undefined; }'.
  Property 'description' does not exist on type 'string & { tags?: FindOperator<ProductTag> | undefined; price_list_id?: FindOperator<PriceList> | undefined; }'.
```

Let me know if we can do anything to make it work nicely with your suggestion 😄

Comment (adrien2p) [reply packages/medusa/src/repositories/product.ts:370]: don't worry about this suggestion 😂 this is because the consumer always return a query from `buildQuery` which is an object and it is passed to the repositories method, but never a string or undefined like `CustomOptions` suggest. it is in fact `Partial<Writable<TEntity>>` but here the type is more generic and include `string | undefined` and some others. So for now it is not very important to get that micro improvement.

**Review by adrien2p — APPROVED** [review]
